Fix bug with printStackTraceToFile() function and make stack printing…

qchaupds opened the pull request:
… optional with new command line parameter --enable-stack-printing

1. Add test resource :src/test/resources/github379
2. Add test resource :src/test/resources/github380 (Directory is required due to quirkiness with behave)
3. Add logic to enable stack printing :src/main/java/gov/nasa/pds/tools/util/FileService.java
4. Add house keeping of stackPrintingFlag variable :src/main/java/gov/nasa/pds/tools/util/FlagsUtil.java
5. Fix bug with printStackTraceToFile(0 function :src/main/java/gov/nasa/pds/tools/validate/rule/pds4/TableDataContentValidationRule.java
6. Add parsing ENABLE_STACK_PRINTING parameter :src/main/java/gov/nasa/pds/validate/ValidateLauncher.java
7. Add ENABLE_STACK_PRINTING parameter :src/main/java/gov/nasa/pds/validate/commandline/options/Flag.java
8. Add ENABLE_STACK_PRINTING option :src/main/java/gov/nasa/pds/validate/commandline/options/FlagOptions.java
9. Add tests for github379 and github380 :src/test/resources/features/validate.feature

Refs:

https://github.com/nasa-pds/validate/issues/379 FileService:printStackTraceToFile:ERROR when validating a product with overlapping fields
https://github.com/nasa-pds/validate/issues/380 stack trace being created during successful validate execution

Closes:

https://github.com/nasa-pds/validate/issues/379 FileService:printStackTraceToFile:ERROR when validating a product with overlapping fields  
https://github.com/nasa-pds/validate/issues/380 stack trace being created during successful validate execution

Tested in DEV:

For issue #379: There should be no crashing.  The report file will properly report there was an issue with 
```
"error.table.field_value_overlap"
              "severity": "ERROR",
              "type": "error.table.field_value_overlap",

% validate -R pds4.label --skip-context-reference-check -r report_github379_label_invalid.json   -s json -t  src/test/resources/github379/mix_cal_hk_fpac_report_20181204.xml  
```
For issue #380: There should be no crashing.  The report file will properly report there was an issue with
```
"error.table.field_value_overlap"
              "severity": "ERROR",
              "type": "error.table.field_value_overlap",

# This test re-use resources from github379
% validate -R pds4.label --skip-context-validation --skip-context-reference-check -r report_github380_label_invalid_without_stack.json   -s json -t  src/test/resources/github379/mix_cal_hk_fpac_report_20181204.xml  >& t2
```

If the --debug-mode is used, the stack will be written to a stack file with the name validate_stack_traces.log
Make sure the file is delete before running next test
```
% rm validate_stack_traces.log
% validate -R pds4.label --debug-mode --skip-context-validation --skip-context-reference-check -r report_github380_label_invalid_with_stack.json   -s json -t  src/test/resources/github380/mix_cal_hk_fpac_report_20181204.xml  >& t3
```

Commit:
Fix bug with printStackTraceToFile() function and make stack printing optional with new command line parameter --enable-stack-printing

1. Add test resource :src/test/resources/github379
2. Add test resource :src/test/resources/github380 (Directory is required due to quirkiness with behave)
3. Add logic to enable stack printing :src/main/java/gov/nasa/pds/tools/util/FileService.java
4. Add house keeping of stackPrintingFlag variable :src/main/java/gov/nasa/pds/tools/util/FlagsUtil.java
5. Fix bug with printStackTraceToFile(0 function :src/main/java/gov/nasa/pds/tools/validate/rule/pds4/TableDataContentValidationRule.java
6. Add parsing ENABLE_STACK_PRINTING parameter :src/main/java/gov/nasa/pds/validate/ValidateLauncher.java
7. Add ENABLE_STACK_PRINTING parameter :src/main/java/gov/nasa/pds/validate/commandline/options/Flag.java
8. Add ENABLE_STACK_PRINTING option :src/main/java/gov/nasa/pds/validate/commandline/options/FlagOptions.java
9. Add tests for github379 and github380 :src/test/resources/features/validate.feature

Refs:

https://github.com/nasa-pds/validate/issues/379 FileService:printStackTraceToFile:ERROR when validating a product with overlapping fields
https://github.com/nasa-pds/validate/issues/380 stack trace being created during successful validate execution

jordanpadams commented on src/main/java/gov/nasa/pds/validate/commandline/options/Flag.java:
```diff
@@ -171,6 +171,8 @@
     // Flag to allow user to check in between fields for non-blank characters for Table_Character validation.
     CHECK_INBETWEEN_FIELDS(null, "strict-field-checks", "Specific to character tables (Table_Character) validation, enable checks to ensure no unexpected alphanumeric characters appear in between in between fields."),
 
+    ENABLE_STACK_PRINTING(null, "enable-stack-printing", "Enable stack trace printing to an external log file if an unexpected exception is caught."),
```
@qchaupds apologies for not providing this input from the beginning, but now that I am seeing this, in order to avoid confusion with the users, can we change this argument to be `--debug-mode`? And when enabled, it will both (1) print stack trace, (2) sets the verbosity to 0 (per our email thread, this is that "hidden" feature that we can unhide using this flag). this will allow us to add new "debug" features in the futures if we want (e.g. email us the stack trace)

qchaupds replied:
@jordanpadams With the following parameter --debug-mode validate now print the stack trace log

{pds-dev3.jpl.nasa.gov}/home/qchau/sandbox/validate 116 % validate -R pds4.label --debug-mode --skip-context-validation --skip-context-reference-check -r report_github380_label_invalid_with_stack.json   -s json -t  src/test/resources/github379/mix_cal_hk_fpac_report_20181204.xml  >& t3

{pds-dev3.jpl.nasa.gov}/home/qchau/sandbox/validate 120 % ls -l *.log
-rw-r--r-- 1 qchau pds 1670 Aug 12 09:52 validate_stack_traces.log

Commit:
Modify --enable-stack-printing to --debug-mode and set severity to 0

1. Add setSeverity() to 0 when --debug-mode is used at command line :src/main/java/gov/nasa/pds/validate/ValidateLauncher.java
2. Modify --enable-stack-printing to --debug-mode :src/main/java/gov/nasa/pds/validate/commandline/options/FlagOptions.java
3. Modify --enable-stack-printing to --debug-mode :src/test/resources/features/validate.feature

Refs:

https://github.com/nasa-pds/validate/issues/379 FileService:printStackTraceToFile:ERROR when validating a product with overlapping fields
https://github.com/nasa-pds/validate/issues/380 stack trace being created during successful validate execution

jordanpadams committed:
Merge branch 'main' into #379_380
